pkg/kstatus/watcher/object_status_reporter.go

```diff
@@ -0,0 +1,872 @@
+		w.stopInformer(gkn)
+	})
+}
+
+// newStatusCheckTaskFunc returns a taskFund that reads the status of an object
+// from the cluster and sends it over the event channel.
+//
+// This method should only be used for generated resource objects, as it's much
+// slower at scale than watching the resource for updates.
+func (w *ObjectStatusReporter) newStatusCheckTaskFunc(
+	ctx context.Context,
+	eventCh chan<- event.Event,
+	id object.ObjMetadata,
+) taskFunc {
+	return func() {
+		klog.V(5).Infof("Re-reading object status: %s", status.ScheduleWindow, id)
+		// check again
+		rs, err := w.readStatusFromCluster(ctx, id)
+		if err != nil {
+			// Send error event and stop the reporter!
+			// TODO: retry N times before terminating
+			w.handleTerminalError(eventCh, err)
+			return
+		}
+		eventCh <- event.Event{
+			Type:     event.ResourceUpdateEvent,
+			Resource: rs,
+		}
+	}
+}
+
+func (w *ObjectStatusReporter) handleTerminalError(eventCh chan<- event.Event, err error) {
+	if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+		klog.V(5).Infof("Watch closed: %v", err)
+		return
+	}
+	eventCh <- event.Event{
+		Type:  event.ErrorEvent,
+		Error: err,
+	}
+	// Terminate the reporter.
+	w.cancel()
+}
+
+// watchErrorHandler logs errors and cancels the informer for this GroupKind
+// if the NotFound error is received, which usually means the CRD was deleted.
+// Based on DefaultWatchErrorHandler from k8s.io/client-go@v0.23.2/tools/cache/reflector.go
+func (w *ObjectStatusReporter) watchErrorHandler(gkn GroupKindNamespace, err error) {
+	// Note: Informers use a stop channel, not a Context, so we don't expect
+	// Canceled or DeadlineExceeded here.
+	switch {
+	case err == io.EOF:
+		// Stop channel closed
+		klog.V(5).Infof("Watch closed: %v: %v", gkn, err)
+	case err == io.ErrUnexpectedEOF:
+		// Keep retrying
+		klog.V(1).Infof("Watch closed: %v: %v", gkn, err)
+	case errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded):
+		// Context cancelled
+		klog.V(5).Infof("Watch closed: %v: %v", gkn, err)
+	case apierrors.IsResourceExpired(err): // resourceVersion too old
+		// Keep retrying
+		klog.V(5).Infof("Watch closed: %v: %v", gkn, err)
+	case apierrors.IsGone(err): // DEPRECATED
+		// Keep retrying
+		klog.V(5).Infof("Watch closed: %v: %v", gkn, err)
+	case apierrors.IsNotFound(err) || containsNotFoundMessage(err): // CRD deleted or not created
+		// Stop watching this resource
+		klog.V(3).Infof("Watch error: %v: stopping all watchers for this GroupKind: %v", gkn, err)
+
+		// Stop all informers for this GK
+		w.forEachTargetWithGroupKind(gkn.GroupKind(), func(gkn GroupKindNamespace) {
+			w.stopInformer(gkn)
+		})
+	default:
+		// Keep retrying
+		klog.Warningf("Watch error (will retry): %v: %v", gkn, err)
+	}
+}
+
+// resourceNotFoundMessage is the condition message for metav1.StatusReasonNotFound.
+const resourceNotFoundMessage = "the server could not find the requested resource"
```

Comment:
Would be nice to have a test that fails if/when the message is changed upstream.

Reply:
I added a test for both List and Watch failure. Interestingly, the Watch code seems to correctly wrap errors, so `apierrors.IsNotFound(err)` works for that, but the List code wraps with a `%v`, not a `%w`, so we have to use both methods in the error handler.

FWIW, if a CRD is deleted while a Watch is executing, it looks like the Watch connection just closes, which will cause the ListAndWatch to return and retry. Then the List will error on the retry, triggering the Informer to be stopped by the ObjectStatusReporter error handler.